compiler/resolution/preFold.cpp

```diff
@@ -458,6 +460,140 @@ static void setRecordDefaultValueFlags(AggregateType* at) {
   }
 }
 
+//
+// Infer the correct storage kind for a context manager resource when no
+// explicit storage is given, e.g.
+//
+//    manage myManager() as myResource do ...;
+//
+// Look for PRIM_INIT_VAR calls where the LHS is a variable marked with
+// FLAG_MANAGER_RESOURCE_INFER_STORAGE, then chase the call temp on
+// the RHS (it should always be a call temp).
+//
+// If the call temp was initialized by a ContextCall, search through the CC
+// and record if:
+//
+//    - Any ref overload is present (if so, we will prefer them)
+//    - If both ref/const ref overloads are present
+//    - If const ref is present but not ref
+//
+// If any ref overload is present, we replace the PRIM_INIT_VAR with a
+// PRIM_MOVE, and mark the LHS with FLAG_REF_VAR.
+//
+// If both ref/const ref overloads are present, we also mark the LHS with
+// FLAG_REF_IF_MODIFIED, because the overload to select depends on how
+// the resource is used.
+//
+// Lastly, if only the const ref overload is present (but not the ref), we
+// always prefer it over a value overload.
+//
+// If only the value overload exists, leave the PRIM_INIT_VAR call alone.
+// The init= call it gets lowered to will be elided later regardless.
+//
```

Comment:
Why do we have to do this here? Can't we let cullOverReferences do the job when we get there in compilation?

Reply:
In other words, we keep the `FLAG_MANAGER_RESOURCE_INFER_STORAGE` on the symbol, and make `cullOverReferences` aware of it.

Reply:
I could do this, but I am really hesitant to touch anything in cullOverReferences as it seems like there are some other prominent bugs. It seemed easier to just intercept the context call and patch in the references before allowing the rest of resolution to complete.